core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/UnitOfWorkTransactionContext.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.user.core.util;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.sql.DataSource;
+
+/**
+ * Support class to maintain a mapping between dataSource and connection. At the same time it maintains commit and
+ * rollback transaction.
+ * By calling commitAllConnection() method user can commit all the transaction. If something happen at the middle
+ * then it will rollback the remaining connection.
+ * By calling rollbackAllConnection() method user can rollback all the transaction. If something happen at the middle
+ * then it will ignore the current transaction and rollback the remaining transaction.
+ * By calling closeConnection() method user can closs all the open connection. If something happen at the middle
+ * then it will ignore the current connection and close the remaining connections.
+ */
+public class UnitOfWorkTransactionContext {
+
+    private static final Log log = LogFactory.getLog(UnitOfWorkTransactionContext.class);
+    private HashMap<DataSource, Connection> activeConnection = new HashMap();
+    private int transactionDepth = 0;
+    private boolean errorOccurred = false;
+
+    /**
+     * Commit all the transaction, if something happened at the middle then it print all the commit and uncommitted
+     * transaction.
+     */
+    void commitAllConnection() {
+
+        StringBuilder commitErrorHandling = new StringBuilder();
```

Comment:
can we create these objects when needed. Usually failure is a rare case. We do not want to create un-necessary objects per each successful transactions, as it will cause more garbage.